### PR TITLE
Improve exception msg by adding column name

### PR DESCRIPTION
**Pull request by wangyems**
For the issue https://github.com/dotnet/machinelearning/issues/5211

In this case, when user choose a big dataset and one or more columns have more than expected unique values, the crossvalidate() method throws overflow but it's difficult for users to realize which column in the dataset cause the problem. If we append column name to the overflow message, user can quickly find where goes wrong.

### CODE REVIEW

File: src/Microsoft.ML.Data/Transforms/KeyToVector.cs [hunk 1]
```diff
@@ -246,14 +246,24 @@ public Mapper(KeyToVectorMappingTransformer parent, DataViewSchema inputSchema)
                 _parent = parent;
                 _infos = CreateInfos(inputSchema);
                 _types = new VectorDataViewType[_parent.ColumnPairs.Length];
-                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                try
```
Comment (harishsk) [inline src/Microsoft.ML.Data/Transforms/KeyToVector.cs:249]: Please add a note here as to why this change is being introduced.

File: src/Microsoft.ML.Data/Transforms/KeyToVector.cs [hunk 2]
```diff
@@ -246,14 +246,24 @@ public Mapper(KeyToVectorMappingTransformer parent, DataViewSchema inputSchema)
...
                 {
-                    int valueCount = _infos[i].TypeSrc.GetValueCount();
-                    int keyCount = _infos[i].TypeSrc.GetItemType().GetKeyCountAsInt32(Host);
-                    if (_parent._columns[i].OutputCountVector || valueCount == 1)
-                        _types[i] = new VectorDataViewType(NumberDataViewType.Single, keyCount);
-                    else
-                        _types[i] = new VectorDataViewType(NumberDataViewType.Single, valueCount, keyCount);
+                    for (int i = 0; i < _parent.ColumnPairs.Length; i++)
```
Comment (harishsk) [inline src/Microsoft.ML.Data/Transforms/KeyToVector.cs:251]: Can you please add a corresponding test for this? And in the test, please mention the bug that prompted this change.

Comment (wangyems) [reply src/Microsoft.ML.Data/Transforms/KeyToVector.cs:251]: Can we put user's dataset into our tests? Since this exception is triggered by specific dataset and  I can not reproduce the same error using a self-designed dataset

File: src/Microsoft.ML.Data/Transforms/KeyToVector.cs (hunk 1 above)
Comment (wangyems) [reply src/Microsoft.ML.Data/Transforms/KeyToVector.cs:249]: sure

File: src/Microsoft.ML.Data/Transforms/KeyToVector.cs (hunk 2 above)
Comment (harishsk) [reply src/Microsoft.ML.Data/Transforms/KeyToVector.cs:251]: If possible, please try to avoid using user data set. From the description it sound like it was because OneHotHashEncoding was applied to the date column with a lot of unique values. Is it not possible to reproduce that? 

---
In reply to: [439126305](https://github.com/dotnet/machinelearning/pull/5232#discussion_r439126305) [](ancestors = 439126305)

File: test/Microsoft.ML.AutoML.Tests/AutoMLFailureTests.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        [Fact]
+        public void CrossValidationOverflowTest()
+        {
+            // This test is introduced for https://github.com/dotnet/machinelearning/issues/5211
+            // that provides users an informational exception message
+            MLContext mlContext = new MLContext(1);
+
+            IDataView trainingDataView = mlContext.Data.LoadFromTextFile<ModelInput>(
+                                            path: GetDataPath("cross_validation_overflow_dataset.txt"),
+                                            hasHeader: true,
```
Comment (harishsk) [inline test/Microsoft.ML.AutoML.Tests/AutoMLFailureTests.cs:40]: Does this need to be stored as a file? Is it not possible to generate random values in the column and reproduce the same issue?

File: test/Microsoft.ML.AutoML.Tests/AutoMLFailureTests.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                                            separatorChar: '\t',
+                                            allowQuoting: true,
+                                            allowSparse: false);
+
+            IDataView testDataView = mlContext.Data.BootstrapSample(trainingDataView);
+
+            ExperimentResult<MulticlassClassificationMetrics> experimentResult = mlContext.Auto()
+                .CreateMulticlassClassificationExperiment(60)
```
Comment (harishsk) [inline test/Microsoft.ML.AutoML.Tests/AutoMLFailureTests.cs:48]: The original repro is from using AutoML API. But the root cause of the bug is due to overflow in VectorDataViewType, Is it possible to constrain the test case to a simpler single pipeline that demonstrates the issue?

File: test/Microsoft.ML.AutoML.Tests/AutoMLFailureTests.cs (hunk 3 above)
Comment (wangyems) [reply test/Microsoft.ML.AutoML.Tests/AutoMLFailureTests.cs:40]: It might be possible, let me try it

File: test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs [hunk 5]
```diff
@@ -222,5 +223,51 @@ public void TestOldSavingAndLoading()
...
+
+
+            [ColumnName("ProblematicColumn"), LoadColumn(1)]
+            public string ProblematicColumn { get; set; }
+        }
+
+        static IEnumerable<ModelInput> GetData()
+        {
+            for (int i = 0; i < 1000; i++)
+            {
```
Comment (harishsk) [inline test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: Are we seeing overflow errors at only 1000 unique values? Does this mean we cannot perform one hot hash encoding on a 1000 unique values?

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: In this case, yes. Can't we perform OneHotHashEncoding on a thousand values? I don't think so.
This pipeline is to simulate how AutoML reacts toward a bigger dataset using a small dataset.
If we merely use OneHotHashEncoding for a thousand unique values there should be no problem(no overflow). However, the pipeline designed by AutoML appends a ColumnCopy pipeline after OneHotHashEncoding which causes the overflow. But normally(My view) we do not do this way.

Comment (harishsk) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: Does the bug repro if we don't have a CopyColumns after OneHotHashEncoding?


---
In reply to: [439639201](https://github.com/dotnet/machinelearning/pull/5232#discussion_r439639201) [](ancestors = 439639201)

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: no

Comment (harishsk) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: That sounds a little suspicious. Can you please dig a little deeper? CopyColumns shouldn't be impacting the performance of OneHotHashEncoding.

---
In reply to: [439641032](https://github.com/dotnet/machinelearning/pull/5232#discussion_r439641032) [](ancestors = 439641032)

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: sure

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: Wait... I'm wrong on your last question. We still can repo the overflow even if without CopyColumns(). So the CopyColumn() is not the reason here. 
Right now the I see the difference is that in AutoML the OneHothashEncoder has been used twice which means:
in normal case:
var result = pipeline.Fit(dataView).Transform(dataView);
in AutoML:
var transformer= pipeline2.Fit(pipeline.Fit(dataView).Transform(dataView)) (pipeline2 and pipeline both have oneHotHashEncoder)

Comment (harishsk) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: Can you please fix the pipeline to the specific conditions that will reproduce the issue? (If CopyColumns is not necessary to reproduce the issue, that shouldn't be part of the pipeline)

---
In reply to: [439646321](https://github.com/dotnet/machinelearning/pull/5232#discussion_r439646321) [](ancestors = 439646321)

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: It's fixed

Comment (harishsk) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: Sorry, I think I was looking at an older iteration. 
This looks good for the ML.NET portion. Do you know why the AutoML.NET code is calling Fit twice?

---
In reply to: [440403606](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440403606) [](ancestors = 440403606)

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: > Sorry, I think I was looking at an older iteration.
> This looks good for the ML.NET portion. Do you know why the AutoML.NET code is calling Fit twice?
> 
> In reply to: [440403606](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440403606) [](ancestors = 440403606)

My understanding is that the first step AutoML does is to find the best way to featurize the original data(this step calls fit and transform), and then feed the model using the featurized data(this step calls fit again)

Comment (harishsk) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: Why does calling Fit twice cause an overflow?

---
In reply to: [440407700](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440407700) [](ancestors = 440407700)

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: As in the second fit(), a Vector<Key<UInt32, 0-65535>, 65536> is required since OneHotHashTransformer is used twice. And that requires a size of 65536*65536 unit of memory which causes overflow

Comment (harishsk) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: The bug occurs not only when you call Fit twice but also if you have two calls to `OneHotHashEncoding` twice in the same pipeline with the same column name.  The behavior of ML.NET is correct in this case.  
However there may be a bug in AutoML.NET in this case because calling it squares the size of the one hot encoded vector and this may be something that is unintended. 

Does AutoML.NET generate code that has the `Fit()` being called twice?





---
In reply to: [440407700](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440407700) [](ancestors = 440407700)

Comment (wangyems) [reply test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:240]: I took a look into our AutoML samples and found that it uses Evaluate() method after selecting the best model. In user's sample however, CrossValidate() is used instead of Evaluate(). Then I realized it's likely that user didn't code correctly since CrossValidate() is a training process. The data is trained twice that's why Fit() been called twice.

We might want to change the name of CrossValidate() to CrossValidateTraining()? As I think it's sort of confusing and let people easily link this with inference process.

File: test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs [hunk 6]
```diff
@@ -222,5 +223,50 @@ public void TestOldSavingAndLoading()
...
+
+            var featurizedData = pipeline.Fit(dataview).Transform(dataview);
+
+            try
+            {
+                var transformer = pipeline.Fit(featurizedData);
+                Assert.True(false);
+            }
+            catch (System.Exception ex)
+            {
```
Comment (harishsk) [inline test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:265]: You can catch the `OverflowException` instead here.

[https://docs.microsoft.com/en-us/dotnet/api/system.overflowexception?view=netcore-3.1](https://docs.microsoft.com/en-us/dotnet/api/system.overflowexception?view=netcore-3.1)

Comment (harishsk) [inline test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs:265]: This exception also happens if you call OneHotHashEncoding twice in your pipeline. Please add a note about it.
